Fix: Pod terminating stuck because of trying to umount not actual mounted dir

mochizuki875 opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Fix the problem Pod terminating stuck because of trying to umount not actual mounted dir.
In addition, I refactored releted methods and functions.

#### Which issue(s) this PR fixes:

Fixes #114546

#### Special notes for your reviewer:

NONE

#### Does this PR introduce a user-facing change?

```release-note
Fix the problem Pod terminating stuck because of trying to umount not actual mounted dir.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

mochizuki875 committed:
check umount result

mochizuki875 commented:
/sig storage

msau42 commented:
/ok-to-test

mochizuki875 commented:
Hi @andyzhangx , thank you for fixing the issue #114736.
Actually, I've come accross the same issue using In-Tree NFS Volume Plugin(#114546) and try to fix it in this PR.
This PR has the same fixes as your PR, with the addition of using common variable names and logic for `Unmount`, `tryUnmount`, and `forceUmount`.
So could you please check this PR?

andyzhangx commented:
> Hi @andyzhangx , thank you for fixing the issue #114736. Actually, I've come accross the same issue using In-Tree NFS Volume Plugin(#114546) and try to fix it in this PR. This PR has the same fixes as your PR, with the addition of using common variable names and logic for `Unmount`, `tryUnmount`, and `forceUmount`. So could you please check this PR?

@mochizuki875 could you point out which line of code fix the original nfs unmount stuck issue in this PR? thanks.

mochizuki875 commented:
@andyzhangx 
Thank you your reply!
When Pod is terminated, NFS VolumePlugin calls `CleanupMountWithForce`.
https://github.com/kubernetes/kubernetes/blob/0b2e54123f233d87c3a132c94d62d0e551bf60ba/pkg/volume/nfs/nfs.go#L294

 And then, calls `UnmountWithForce`.
 https://github.com/kubernetes/kubernetes/blob/0b2e54123f233d87c3a132c94d62d0e551bf60ba/staging/src/k8s.io/mount-utils/mount_helper_common.go#L58

Finally, `tryUnmount` and `forceUmount` are called.
https://github.com/kubernetes/kubernetes/blob/891eb75b8956c94c59dbfab29e78f12863202088/staging/src/k8s.io/mount-utils/mount_linux.go#L373-L383

Before your PR(#114736), there was not the check logic of umount result in `tryUnmount` and `forceUmount`, so an error occured and kubelet could not finish the terminating proccess of Pod.
So I added `checkUmountError` function as a common logic to check a result of `umount` command.
https://github.com/kubernetes/kubernetes/blob/891eb75b8956c94c59dbfab29e78f12863202088/staging/src/k8s.io/mount-utils/mount_linux.go#L796-L811

And I implemented `checkUmountError` to `Unmount`, `tryUnmount`, and `forceUmount` because they need same logic to check a result of `umount` command.

https://github.com/kubernetes/kubernetes/blob/891eb75b8956c94c59dbfab29e78f12863202088/staging/src/k8s.io/mount-utils/mount_linux.go#L361-L369
https://github.com/kubernetes/kubernetes/blob/891eb75b8956c94c59dbfab29e78f12863202088/staging/src/k8s.io/mount-utils/mount_linux.go#L766-L794

Also, although it is a small detail, the names of variables that indicate the same meaning were different between `Unmount`, `tryUnmount`, and `forceUmount`(ex. `target` and `path`, `command` and `cmd`), so I corrected to common names.

bart0sh commented:
/priority important-soon
/triage accepted

xing-yang commented:
/assign @gnufied

gnufied commented:
Can you rebase this PR please?

mochizuki875 commented:
/retest

andyzhangx commented:
> @andyzhangx Thank you for your reply! When Pod is terminated, NFS VolumePlugin calls `CleanupMountWithForce`.
> 
> https://github.com/kubernetes/kubernetes/blob/0b2e54123f233d87c3a132c94d62d0e551bf60ba/pkg/volume/nfs/nfs.go#L294
> 
> And then, calls `UnmountWithForce`.
> 
> https://github.com/kubernetes/kubernetes/blob/0b2e54123f233d87c3a132c94d62d0e551bf60ba/staging/src/k8s.io/mount-utils/mount_helper_common.go#L58
> 
> Finally, `tryUnmount` and `forceUmount` are called.
> 
> https://github.com/kubernetes/kubernetes/blob/891eb75b8956c94c59dbfab29e78f12863202088/staging/src/k8s.io/mount-utils/mount_linux.go#L373-L383
> 
> Before your PR(#114736), there was not the check logic of umount result in `tryUnmount` and `forceUmount`, so an error occured and kubelet could not finish the terminating proccess of Pod if NFS volume is not mounted to Pod.(In this case, umount command returns "not mounted".) So I added `checkUmountError` function as a common logic to check a result of `umount` command.
> 
> https://github.com/kubernetes/kubernetes/blob/891eb75b8956c94c59dbfab29e78f12863202088/staging/src/k8s.io/mount-utils/mount_linux.go#L796-L811
> 
> And I implemented `checkUmountError` to `Unmount`, `tryUnmount`, and `forceUmount` because they need same logic to check a result of `umount` command.
> 
> https://github.com/kubernetes/kubernetes/blob/891eb75b8956c94c59dbfab29e78f12863202088/staging/src/k8s.io/mount-utils/mount_linux.go#L361-L369
> 
> 
> https://github.com/kubernetes/kubernetes/blob/891eb75b8956c94c59dbfab29e78f12863202088/staging/src/k8s.io/mount-utils/mount_linux.go#L766-L794
> 
> Also, although it is a small detail, the names of variables that indicate the same meaning were different between `Unmount`, `tryUnmount`, and `forceUmount`(ex. `target` and `path`, `command` and `cmd`), so I corrected to common names.

@mochizuki875 so this https://github.com/kubernetes/kubernetes/pull/114736 already fixed the issue since it added following code into the `tryUnmount` func, and your PR is more like a refactor and cleanup, right?
```
		if mounter.withSafeNotMountedBehavior && strings.Contains(string(out), errNotMounted) {
			klog.V(4).Infof("ignoring 'not mounted' error for %s", path)
			return nil
		}
```

mochizuki875 commented:
> so this https://github.com/kubernetes/kubernetes/pull/114736 already fixed the umount not actual mounted dir issue since it added following code into the tryUnmount func, and your PR is more like a refactor and cleanup, right?

@andyzhangx 
Yes, in addition, I implemented the same check logic to `forceUmount`.
If `tryUnmount` fails in `UnmountWithForce`, then `forceUmount`(`umount -f`) will be called as the next step, so the same logic seems to be required.

andyzhangx reviewed (approved):
/lgtm

andyzhangx commented:
the fix code lgtm, is there any way to write a unit test for this logic?

andyzhangx commented:
/test pull-kubernetes-e2e-aks-engine-azure-file

andyzhangx commented:
/test pull-kubernetes-e2e-capz-azure-disk
/test pull-kubernetes-e2e-capz-azure-disk-vmss
/test pull-kubernetes-e2e-capz-azure-file
/test pull-kubernetes-e2e-capz-azure-file-vmss

mochizuki875 commented:
> the fix code lgtm
thx!

> is there any way to write a unit test for this logic?
@andyzhangx 
Sorry, I don't know right away...
It might need to emulate a mountpoint to check the result of umount command in unit test case, but I'm not sure how to do that.

gnufied commented:
I think the fix is reasonable.

/lgtm

gnufied commented:
We should add a release note. Because this is a user visible change afaict.

mochizuki875 commented:
> We should add a release note. Because this is a user visible change afaict.

@gnufied 
I added release node. Please check it.

mochizuki875 committed:
add unit test for check unmounted behavior of Unmount

mochizuki875 commented:
> is there any way to write a unit test for this logic?

@andyzhangx 
I tried to write a unit test code like this.
This code checks the behavior of `Unmount`, `tryUnmount`, and `forceUmount` to not mounted dir.

https://github.com/kubernetes/kubernetes/blob/41987b975b7be2f4da4b7c2c7027506d66a024dd/staging/src/k8s.io/mount-utils/mount_linux_test.go#L626-L655

mochizuki875 commented:
/retest

mochizuki875 commented:
/retest

mochizuki875 commented:
/retest

mochizuki875 commented:
@andyzhangx @gnufied
PTAL?

gnufied commented:
/lgtm

/assign @jsafrane

jsafrane commented:
/approve

jsafrane commented on staging/src/k8s.io/mount-utils/mount_linux_test.go:
```diff
@@ -620,3 +621,28 @@ func makeFakeCommandAction(stdout string, err error) testexec.FakeCommandAction
 	}
 }
+
+func TestNotMountedBehaviorOfUnmount(t *testing.T) {
```
Opened a revert: https://github.com/kubernetes/kubernetes/pull/115732

mochizuki875 replied:
I am terribly sorry...
Please wait a moment.
I'll try to fix it.

saurav-agarwalla commented:
Are there any plans to backport this to older Kubernetes versions? This seems like a critical fix.
